### PR TITLE
fix(material/input): Do not set `aria-invalid` on empty`matInput`s

### LINKED ISSUE #22777: bug(matInput): Screen Readers do not announce that an empty `matInput[required]` is invalid
#### Reproduction
StackBlitz: https://components-issue-pcy3uq.stackblitz.io

Steps to reproduce:
1. Turn on a screen reader
2. Focus the input
3. Move focus away from the input
4. Return focus to the input

#### Expected Behavior
Screen reader announces to the user that the form field is invalid.

#### Actual Behavior
Screen reader does not announce to the user that the form field is invalid.

Note that while the value in `mat-error` is announced to the user when the field is focused there is no indication that the field itself is invalid (unless the error message itself contains that).

#### Environment
- Angular:^12.0.1
- CDK/Material: ^12.0.1
- Browser(s): Any
- Operating System (e.g. Windows, macOS, Ubuntu): Any

#### Details
This issue is rooted in the fact that `matInput` sets `aria-invalid="false"` for empty inputs.  This causes screen readers not to announce that a `required` input is invalid despite being empty.

We should update `matInput` to not apply `aria-invalid` on instances that do not have any value within the input.  This may be accomplished via changing [this line within the matInput directive](https://github.com/angular/components/blob/master/src/material/input/input.ts#L86) from:
```
'[attr.aria-invalid]': 'errorState && !empty'
```
to:
```
'[attr.aria-invalid]': 'empty ? null : errorState'
```

Doing this will allow us to comply with [the ARIA spec for `aria-invalid`](https://www.w3.org/TR/wai-aria/#aria-invalid) (since `aria-invalid` isn't applied to the element until the user has entered some data) while also providing the correct announcements (since `aria-required`/`input[required]` is indicating to the screen reader that the empty field is invalid).

##### Testing methodology

I used this HTML document for testing ARIA attribute behavior with VoiceOver on Chrome and JAWS on Chrome:
```
<html>
  <body>
    <form>
      <input required/>
      <button type="submit">Submit</button>
    </form>
  </body>
</html>
```
Testing with adding/removing various aria attributes on the `input` element showed the following behavior:

| ARIA attributes | value | JAWS+Chrome | VoiceOver+Chrome |
|------------|-------|-------------|------------------|
| `aria-required` not present, `aria-invalid` not present| no value | "Edit Required invalid entry" | "required, invalid data, edit text"|
| `aria-required` not present, `aria-invalid` not present| "value" | "Edit Required, value" | "value, required, edit text"|
| `aria-required="true"`, `aria-invalid` not present| no value | "Edit Required, invalid entry" | "required, invalid data, edit text" |
| `aria-required="true"`, `aria-invalid` not present | "value" | "Edit required, value" | "value, required, edit text" |
| `aria-required="true"`, `aria-invalid="true"` | no value | "Edit required, invalid entry" | "required, invalid data, edit text" |
| `aria-required="true"`, `aria-invalid="true"` | "value" | "Edit required, invalid entry" | "value, required, invalid data, edit text" |
| `aria-required="true"`, `aria-invalid="false"` | no value | "Edit required, type in text" | "required, edit text" |
| `aria-required="true"`, `aria-invalid="false"` | "value" | "Edit required, type in text" | "value, required, edit text" |

While I did not test with NVDA/ChromeVox or on FF/IE/etc. I'm reasonably confident that these results will replicate across these use cases as well.

### PR DESCRIPTION
**Pull request by ByzantineFailure**
Updates the logic for setting `aria-invalid` on `matInput` to not set
the attribute at all if the input is `required` and has no value.

Prior to this PR `matInput` sets `aria-invalid="false"` for any empty
`matInput`, including `required` ones.  This suppresses screen readers'
announcement to users that such inputs are in an invalid state.

Fixes #22777

### CODE REVIEW

**Comment by ByzantineFailure** [comment]
This is actually just a continuation of #22779 but I screwed up the branch pretty severely when I:
A) squash merged it (really not the root of the problem)
B) pushed the new branch name I had on the local since I messed up the name on my remote (the actual problem)

There is good discussion on that PR which needs to be followed up on.  Copying over: 

Via @crisbeto : This behavior is intentional, see #21609. There's also a linked issue.

Response:
Yep! The reason #21609 was requested was because without it aria-invalid was being applied to aria-required inputs with no value. This violates the ARIA spec (see here : https://www.w3.org/TR/wai-aria/#aria-invalid ) and caused #18140 to be filed.

Unfortunately, setting aria-invalid="false" on these inputs suppresses any native indication to a screen reader user that the input is in an invalid state. The only information they receive is the error message itself. If worded correctly the error message can be enough (e.g. if it includes "Invalid" in the text) but this requires component users to be aware of this fact and actively account for it.

Without a well-worded error message, this results in a different experience for SR users than visual users -- the input is flagged as invalid and displayed prominently to visual users as invalid, but screen reader users get no native notification at all when they focus the input.

File: src/material-experimental/mdc-input/input.spec.ts [hunk 1]
```diff
@@ -1081,7 +1081,7 @@ describe('MatMdcInput with forms', () => {
       expect(describedBy).toBe(errorIds);
     }));
 
-    it('should not set `aria-invalid` to true if the input is empty', fakeAsync(() => {
+    it('should set `aria-invalid` to true if the input is empty', fakeAsync(() => {
```
Comment (ByzantineFailure) [inline src/material-experimental/mdc-input/input.spec.ts:1084]: Note:  This test checks that a non-`required` `input` with a `FormControl` which contains `Validators.required` sets `aria-invalid`.

This actually is the edge case that @zelliott pointed out in #22779 when they mentioned that the logic for setting `aria-invalid` should actually be `(empty && required) ? null : errorState`.

In this case there is no `required` on the form field, and therefore no `aria-required`.  This means we _must_ set `aria-invalid` to indicate to a screen reader that there is some other validation error on the form field that they need to be aware of.

I'm not certain that this test's title is informative enough in this case -- thoughts?

File: src/material-experimental/mdc-input/input.spec.ts [hunk 2]
```diff
@@ -1823,3 +1865,14 @@ class MatInputWithColor {
   `
 })
 class MatInputInsideOutsideFormField {}
+
+@Component({
+  template: `
+    <mat-form-field>
+      <input matInput required [formControl]="formControl">
```
Comment (crisbeto) [inline src/material-experimental/mdc-input/input.spec.ts:1872]: I think that the `ReactiveFormsModule` discourages using `required` (and similar attributes) together with `formControl`, in favor of passing in `Validator` functions. That being said, we currently have an issue where we can't detect `Validators.required` from a `FormControl` because `@angular/forms` doesn't have an API for it. Are these tests capturing something that isn't captured by the other `ngModel`-based tests?

Comment (ByzantineFailure) [reply src/material-experimental/mdc-input/input.spec.ts:1872]: Assuming I'm reading the tests correctly, what exists right now does a good job of testing a `required` `matInput` and a good job of testing a `matInput` which relies on the `ReactiveFormsModule` without the native `required` attribute.

These specifically test when both are present since the new logic considers both when assigning `aria-required`.  If we are okay with leaving that scenario untested since it's discouraged by `ReactiveFormsModule` I'm fine to remove these tests.  I included them for completeness' sake more than anything.

Comment (crisbeto) [reply src/material-experimental/mdc-input/input.spec.ts:1872]: These `ReactiveFormsModule` tests can safely be removed since they don't reflect how the component should be used.

Comment (ByzantineFailure) [reply src/material-experimental/mdc-input/input.spec.ts]: These tests have been removed.

**Review by crisbeto — APPROVED** [review]
LGTM
